.github/workflows/main.yml

```
@@ -1,193 +1,135 @@
-        miniforge-version: latest
-        activate-environment: cpp_test
-        environment-file: tests/ci_build/conda_env/cpp_test.yml
-        use-mamba: true
-    - name: Display Conda env
-      run: |
-        conda info
-        conda list
-
-    - name: Build and install XGBoost static library
-      run: |
-        mkdir build
-        cd build
-        cmake .. -DBUILD_STATIC_LIB=ON -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -GNinja
-        ninja -v install
-        cd -
-    - name: Build and run C API demo with static
-      run: |
-        pushd .
-        cd demo/c-api/
-        mkdir build
-        cd build
-        cmake .. -GNinja -DCMAKE_PREFIX_PATH=$CONDA_PREFIX
-        ninja -v
-        ctest
-        cd ..
-        rm -rf ./build
-        popd
-
-    - name: Build and install XGBoost shared library
-      run: |
-        cd build
-        cmake .. -DBUILD_STATIC_LIB=OFF -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX -GNinja -DPLUGIN_FEDERATED=ON -DGOOGLE_TEST=ON
-        ninja -v install
-        ./testxgboost
-        cd -
-    - name: Build and run C API demo with shared
-      run: |
-        pushd .
-        cd demo/c-api/
-        mkdir build
-        cd build
-        cmake .. -GNinja -DCMAKE_PREFIX_PATH=$CONDA_PREFIX
-        ninja -v
-        ctest
-        popd
-        ./tests/ci_build/verify_link.sh ./demo/c-api/build/basic/api-demo
-        ./tests/ci_build/verify_link.sh ./demo/c-api/build/external-memory/external-memory-demo
-
-  cpp-lint:
-    runs-on: ubuntu-latest
-    name: Code linting for C++
+        include:
+          - description: single-gpu
+            container: xgb-ci.gpu
+            suite: gpu
+            runner: linux-amd64-gpu
+            artifact_from: build-cuda
+          - description: multiple-gpu
+            container: xgb-ci.gpu
+            suite: mgpu
+            runner: linux-amd64-mgpu
+            artifact_from: build-cuda
     steps:
-    - uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6
-      with:
-        submodules: 'true'
-    - uses: actions/setup-python@0b93645e9fea7318ecaed2b359559ac225c90a2b # v5.3.0
-      with:
-        python-version: "3.10"
-        architecture: 'x64'
-    - name: Install Python packages
-      run: |
-        python -m pip install wheel setuptools cmakelint cpplint==1.6.1 pylint
-    - name: Run lint
-      run: |
-        python3 tests/ci_build/lint_cpp.py
-        sh ./tests/ci_build/lint_cmake.sh
+      # Restart Docker daemon so that it recognizes the ephemeral disks
+      - run: sudo systemctl restart docker
+      - uses: actions/checkout@v4
+        with:
+          submodules: "true"
+      - name: Fetch container from cache
+        run: bash ops/docker_build.sh ${{ matrix.container }}
+      - name: Unstash Python wheel
+        run: |
+          bash ops/pipeline/stash-artifacts.sh unstash ${{ matrix.artifact_from }} \
+            python-package/dist/*.whl ./xgboost
+          chmod +x ./xgboost
```

```suggestion
          chmod +x ./xgboost
```

Is `./xgboost` the `xgboost` CLI? If so, why is it being download and modified here in a CI job that appears to only be running Python tests?

And if not.... what is it?

---

Yes, `./xgboost` is a CLI executable. It is needed here because some of the pytests test CLI examples. See https://github.com/dmlc/xgboost/blob/54930ec2e374c460c3ba3c44c1da850afc0f81f8/tests/python/test_demos.py#L220.

---

Ah, I didn't expect that. Ok thanks for clarifying, makes sense.